tests/phpunit/tests/option/option.php

```diff
@@ -279,6 +279,113 @@ public function test_special_option_name_notoptions() {
 		delete_option( 'notoptions' );
 	}
 
+	/**
+	 * Test cases for testing whether update_option() will add a non-existent option.
+	 */
+	public function data_option_values() {
+		return array(
+			array( '1' ),
+			array( 1 ),
+			array( 1.0 ),
+			array( true ),
+			array( 'true' ),
+			array( '0' ),
+			array( 0 ),
+			array( 0.0 ),
+			array( false ),
+			array( '' ),
+			array( null ),
+			array( array() ),
+		);
+	}
+
+	/**
+	 * Tests that a non-existent option is added only when the pre-filter matches the default 'false'.
+	 *
+	 * @ticket 22192
+	 * @dataProvider data_option_values
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_with_false_pre_filter_adds_missing_option( $option ) {
+		// Filter the old option value to `false`.
+		add_filter( 'pre_option_foo', '__return_false' );
+
+		/*
+		 * When the option is equal to the filtered version, update option will bail early.
+		 * Otherwise, The pre-filter will make the old option `false`, which is equal to the
+		 *  default value. This causes an add_option() to be triggered.
+		 */
+		if ( false === $option ) {
+			$this->assertFalse( update_option( 'foo', $option ) );
+		} else {
+			$this->assertTrue( update_option( 'foo', $option ) );
+		}
+	}
+
+	/**
+	 * Tests that a non-existent option is never added when the pre-filter is not 'false'.
+	 *
+	 * @ticket 22192
+	 * @dataProvider data_option_values
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_with_truthy_pre_filter_does_not_add_missing_option( $option ) {
+		// Filter the old option value to `true`.
+		add_filter( 'pre_option_foo', '__return_true' );
+
+		$this->assertFalse( update_option( 'foo', $option ) );
+	}
```

Comment:
I'm not sure I understand this test. Can you clarify why the option shouldn't be updated here? Except for the value of `true`, I would expect this to succeed here throughout.

Reply:
It’s not about whether it _should_ be. It’s about whether the current behavior does. 

Essentially, what I think is happening is that when the old value is filtered to be anything other than false, the add_option logic is skipped and it tries to run an update query, but that query fails since the option doesn’t exist in the DB.

Reply:
That makes sense @joemcgill.

To clarify, "when the old value is filtered to be anything other than false" - that's false or the otherwise filtered default value, is that right?

Reply:
I believe so. Yes